rhea/cores/fifo/fifo_fast.py

```diff
@@ -97,7 +97,10 @@ def rtl_fifo():
 
         elif fbus.write and not fbus.read:
             fbus.empty.next = False
-            if not fbus.empty:
+            # because addr starts from 0,
+            # there should not be an increment
+            # the first time 
+            if (not fbus.empty) and (not fbus.full):
```

Comment:
The previous tests were perfectly valid. Since we start the empty queue at addr 0 and not -1, we would need to stay one step behind in the increments ,so the 
```python
if (not fbus.empty) condition is needed
```
I added the other 
```python
if (not fbus.full)
```
because in other instances the fifo continues to increment the `addr` counter even after the fifo is full. This is sometimes prevented because the `nvacant` decrement below zero throws up an exception, but if there were some code, that were to ignore the exception, or keep writing after the fifo is full, `addr` would otherwise end on a totally out of bounds end value, this prevents that. The previous tests you wrote didn't check for out of bounds, and even if they did, they would catch the exception, so this issue doesn't usually occur (It felt semantically more correct too, tell me if you disagree).

If you want me to remove the changes, of course, I will.